### ISSUE TITLE
Release Descriptions contains orphaned records

### ISSUE THREAD

**Issue by miketheman** [pypi/warehouse#14825]
Something about our Release Descriptions tables is causing orphaned records.

This was discovered in #14788 when fixing the `backref` to `back_populates` relationship flags.

This SQL query will expose the orphaned records.
```sql
select count(d.id) from release_descriptions d
left join releases r on d.id = r.description_id
where r.description_id is null;
```
As of this writing, the count is 621643, and had increased by [two in ~4 minutes](https://github.com/pypi/warehouse/pull/14788#discussion_r1376255208). I had thought that the changes over there would prevent records from being orphaned, so something is still causing these records to not be removed.

Since the Descriptions model doesn't maintain a foreign key to it's parent Release, rather the parent Release retains a reference to the current Description, it's not super simple to back-trace these records. We can probably use the `raw` text column to find similar records, but that's no guarantee since many Releases use a pretty similar Description, but it's one idea.

**Comment by miketheman** [comment]
The count today is 666275. There's been changes to the db models since this was observed.

If the number remains stable, then we can entertain a simple `DELETE` for these records. Otherwise we need for fund out how release_description rows are being orphaned.

**Comment by di** [comment]
Looks like it's still in flux:

```
warehouse=> select count(d.id) from release_descriptions d
left join releases r on d.id = r.description_id
where r.description_id is null;
 count
--------
 666374
(1 row)
```

**Comment by miketheman** [comment]
I dug further into this today (we're past 800k).

`releases.description_id` is a FK pointing **up** to `release_descriptions.id` (`ON DELETE CASCADE`). 
That fires when a *description* is deleted (**it removes the release**), never the other way - and that's not something I'd expect, nor do we do as far as I can tell.
There is no FK from `release_descriptions` back to `releases`, so nothing at the database level removes a description when its release goes away, via whatever method at the DB-layer.
The FK relationship originated in #5818.

Cleanup relies on the SQLAlchemy `delete-orphan` cascade on `Release.description`.
That cascade is bypassed whenever a release is removed without the ORM unit of work loading it.
Most importantly when a **project** is deleted: [`Project.releases` uses `passive_deletes=True`](https://github.com/pypi/warehouse/blob/741bd8eb449f6c78138481702f1cbdf74a35e9dc/warehouse/packaging/models.py#L243-L247), so the DB deletes the release rows via the `releases.project_id` `ON DELETE CASCADE` and the descriptions are left orphaned, due to the direction of the FK.

Some ways to fix this:

### A. Add an ORM listener to clean up. Something like this:
```python
@event.listens_for(Project, "before_delete")
def delete_release_descriptions(mapper, connection, target):
    connection.execute(
        delete(Description).where(
            Description.id.in_(
                select(Release.description_id).where(Release.project_id == target.id)
            )
        )
    )
```

That adds yet another ORM-backed listener, which is generally fine, but also adds a little indirection, and does not apply a constraint at the DB level, so there's still the possiblity of some other database operation outside of the warehouse code creating orphans.

### B. Add a DB trigger that does effectively the same as the ORM listener, but at the DB level.
Something like this:
```sql
CREATE OR REPLACE FUNCTION delete_orphaned_release_description()
        RETURNS TRIGGER AS $$
        BEGIN
            DELETE FROM release_descriptions WHERE id = OLD.description_id;
            RETURN NULL;
        END;
        $$ LANGUAGE plpgsql;
```

This has the advantage of being at the DB-level, so any DB-level ops will invoke the trigger, but it's a little less obvious. A big pro here is that the migration takes very little locks, no table rewrites, etc, so it ships quickly (needs `SHARE ROW EXCLUSIVE` lock on `releases`), and should immediately start stemming the drip.

### C. Do the "right thing" - multiple steps

The "correct" declarative fix is to make the description a true child of the release: add `release_descriptions.release_id` with a FK `ON DELETE CASCADE` to `releases`, and retire the `releases.description_id` direction.
This replaces the trigger with native referential integrity.

There's a bunch of steps to get this done right:

1. `ADD COLUMN release_descriptions.release_id uuid` (nullable, fast).
2. Backfill `release_id` from `releases` - full-table `UPDATE` of millions of large-text rows; must be batched.
3. Ensure orphans are gone (step 2 above) before enforcing the constraint.
4. Add the FK `NOT VALID`, then `VALIDATE CONSTRAINT`; add an index on `release_id`.
5. Set `release_id NOT NULL`.
6. Resolve the now-mutual cascade between the two FKs (cycle):
    - drop/rewire the old `releases.description_id` FK
    - update the ORM relationship + any query that reads `description_id`.

Once this lands, the trigger from B can be dropped.

---

I prototyped A to see if it worked, and it does, but I think B is probably better for now, as it also sidesteps any unexpected transaction management quirks we may see in ORM-land, and once that's done and we clean up the orphaned records, we can close this and open a new issue for the steps in C.

**Comment by miketheman** [comment]
Option B shipped in #20216.

Count post-deploy:
```sql
warehouse=> select count(d.id) from release_descriptions d
left join releases r on d.id = r.description_id
where r.description_id is null;
 count
--------
 844170
```
After processing some removals, the count remains stable, so worth checking in a few days to confirm whether this stopped the drip, and then continue with the Option C above.

**Comment by miketheman** [comment]
Checked again today, looks like the fix is holding.

```sql
warehouse=> select count(d.id) from release_descriptions d
left join releases r on d.id = r.description_id
where r.description_id is null;
 count  
--------
 844170
```

The next step is to write a backdrain (opposite of a backfill?) task that removes 5000 of these per run, and run once every 5 minutes, which should succeed in under a day, or more aggressively of that makes sense, but a single day run seems reasonable to me.
